pmlc/dialect/abi/tests/dot.mlir

```diff
@@ -0,0 +1,101 @@
+// RUN: pmlc-opt --lower-to-abi %s | FileCheck %s
+
+module @dot {
+  func @main(%arg0: memref<8x16xf32>, %arg1: memref<16x32xf32>, %arg2: memref<8x32xf32>) {
+    %cst = constant 0.000000e+00 : f32
+    %c0 = constant 0 : index
+    %c1 = constant 1 : index
+    %c2 = constant 2 : index
+    %c3 = constant 3 : index
+    %c5 = constant 5 : index
+    %c6 = constant 6 : index
+    %c8 = constant 8 : index
+    %c7 = constant 7 : index
+    %c4 = constant 4 : index
+    omp.parallel num_threads(%c4 : index) default(shared) {
+      %1 = call @plaidml_rt_thread_num() : () -> index
+      br ^bb1(%c0 : index)
+    ^bb1(%2: index):  // 2 preds: ^bb0, ^bb2
+      %3 = cmpi "slt", %2, %c8 : index
+      cond_br %3, ^bb2, ^bb3
+    ^bb2:  // pred: ^bb1
+      %4 = muli %1, %c8 : index
+      store %cst, %arg2[%2, %4] : memref<8x32xf32>
+      %5 = addi %4, %c1 : index
+      store %cst, %arg2[%2, %5] : memref<8x32xf32>
+      %6 = addi %4, %c2 : index
+      store %cst, %arg2[%2, %6] : memref<8x32xf32>
+      %7 = addi %4, %c3 : index
+      store %cst, %arg2[%2, %7] : memref<8x32xf32>
+      %8 = addi %4, %c4 : index
+      store %cst, %arg2[%2, %8] : memref<8x32xf32>
+      %9 = addi %4, %c5 : index
+      store %cst, %arg2[%2, %9] : memref<8x32xf32>
+      %10 = addi %4, %c6 : index
+      store %cst, %arg2[%2, %10] : memref<8x32xf32>
+      %11 = addi %4, %c7 : index
+      store %cst, %arg2[%2, %11] : memref<8x32xf32>
+      %12 = addi %2, %c1 : index
+      br ^bb1(%12 : index)
+    ^bb3:  // pred: ^bb1
+      omp.terminator
+    }
+    %0 = xsmm.gemm.dispatch.f32 [8, 32, 16], [16, 32, 32]
+    xsmm.gemm.invoke.f32 %0, %arg2[%c0, %c0] = %arg0[%c0, %c0], %arg1[%c0, %c0] : (memref<8x16xf32>, memref<16x32xf32>) -> memref<8x32xf32>
+    return
+  }
+  func @plaidml_rt_thread_num() -> index
+}
+
+// CHECK-LABEL: module @dot {
+//  CHECK-NEXT:   abi.loop init  {
+//  CHECK-NEXT:   ^bb0(%arg0: !llvm.ptr<i8>):  // no predecessors
+//  CHECK-NEXT:     abi.yield
+//  CHECK-NEXT:   } yield [] body  {
```

Comment:
This is confusing, can we just make the syntax be:

```
abi.loop init {
} exec {
} fini {
}
```

Reply:
We need the field types somewhere; it kinda makes sense to put them after the init region.  (There are some hacky alternatives, but it's simplest to just keep the authoritative version in an attribute.)

Reply:
Ok, so:
```
abi.loop init (%arg0 : type, ...) {
} exec (%arg0 : type, ...) {
} fini (%arg0: type, ...) {
}
```

Reply:
It's a little tricky with exec, since there's no clear division between the block arguments coming from the initialized network object and the per-execution block arguments, other than by looking at the networkFieldTypes attribute, which then needs to be stored somewhere.  (We could look at the fini region; that's the hacky alternative I mentioned earlier, but that seems kind of hacky.  An earlier version of this code actually looked at the terminator of the init region, but again, that seemed hacky; I'd rather just be very explicit by having a single authoritative attribute.)

Is there an easy way to use the declarative asm syntax to get nice function-style arguments like this?  If so, I'd love to use it.

Reply:
So I think readability of the IR is very important here.  With regard to the types attribute, the more that I think about it, since the 'networkFieldTypes' is 1-to-1 to the fini params (and a validation error on mismatch), I think we should remove the attribute and instead add an accessor to the op that grabs the fini arg types, which will end up being equally easy to use in the code, and doesn't duplicate data that is required to match anyway.  With regard to the textual format, the ODS assembly is somewhat poorly documented, and so I'd try it out and see what it can do.  I believe that it might just be:
let assemblyFormat = [{ `init` $initRegion  `body` $bodyRegion `fini` $finiRegion attr-dict }];
If that doesn't work (i.e. formats argument anonyingly, has ^bb goo, etc), I'd mess around a bit to try to fix it, and if you can't work around it, write a full custom parser/printer.  Basically, the default format is very annoying to read/write, and it's worth a few hours of pain writing parser/printer logic to fix it.

Reply:
The current impl uses ODS assembly.  :-). I don't think I have the time right now to spend a few hours writing the parser/printer logic for something beyond what ODS can do.  But replacing networkFieldTypes with the "use the fini types" hack is pretty straightforward; I'll go ahead and implement it.

Reply:
There are already traits that allow the ODS assembly format to be smarter by saying things like "these operand types must match those result types", so perhaps we can use some existing traits for this.

Reply:
Once the networkFieldTypes bit is in I will fix the formatting on this branch.